File: week1/largest-perimeter-triangle.py
class Solution:
    def largestPerimeter(self, nums: List[int]) -> int:
        # first sort the array
        # start from the last index and check for validity of three lengths at last
        # iterate until the 3rd index, i = 2
        # if valid return the length of the three numbers
        # else 
        # 
        # return 0
        nums.sort()
        index = len(nums) - 1

        def check_validity(a, b, c):
            return a + b > c and b + c > a and a + c > b

        while index > 1:
            if check_validity(nums[index], nums[index-1], nums[index-2]):
                return nums[index] + nums[index-1] + nums[index-2]
            index -= 1
        return 0
# a + b > c
# a + c > b
# b + c > a

File: leetcode/largest-perimeter-triangle.cpp
class Solution {
public:
    bool isValid(int a, int b, int c)
    {
        return a + b > c  && a + c > b && b + c > a;
    }
    int largestPerimeter(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        for (int i = nums.size() - 1; i >= 2; i--)
        {
            if (isValid(nums[i], nums[i-1], nums[i-2]))
            {
                return nums[i] + nums[i-1] + nums[i-2];
            }
        }
        return 0;
    }
};
// 10 2 1 1
